Chain setup API / Centralize default HF / Common options dict / Block HFbyBlockNumber & initWithGenesisHeader options

holgerd77 opened the pull request:
This PR prepares for the rework of EIP support as a follow-up on #856 

I moved Common to be instantiated with an option dictionary - like we do in most of the other libraries too - should be a good occasion to do this now since we are right before a breaking release anyhow.

This prepares for a new to be added `eips` option, things would otherwise slowly getting too ugly with all the optional parameters chained. It also makes the library more future proof by preparing the ground for future non-breaking option additions.

First push here to see if all tests in related consuming libraries pass.

jochem-brouwer commented on packages/common/src/index.ts:
```diff
@@ -125,18 +140,37 @@ export default class Common {
+   * @param blockNumber
+   * @returns The name of the HF set
+   */
+  setHardforkByBlockNumber(blockNumber: number): string {
```
I think we should also add the possibility that `blockNumber` is `null`. For instance, in all testnet chain configs the DAO support is in the list of supported forks, but the "activation block number" is `null`, i.e. it is never activated (the alternative is to use a super high block number but this is not clean).

jochem-brouwer replied:
Never mind got the semantics of this function wrong.

holgerd77 commented:
Ok, this got a bit out of hand and one thing lead to another. 😄  I will repurpose the target of this PR to be just preparatory work, otherwise things would get too hard too review, hope I am not yet too much out of the scope here. Nevertheless once tests pass here this would be open for review, you should likely do commit-by-commit for a better overview, the single commits are relatively well structured.

This PR addresses a lot of issues we had discussed lately with the `Common` library and touches also some other issues:

- Already mentioned above: `Common` switches to an `options` dictionary for instantiation, 14bc686976b6a1a9f022dad090338da433651319
- The `chain` and `hardfork` constructor options have been removed on all libraries and passing in a `Common` instance is now the single way to adopt the chain setup, see c1bf90eb79c719cb5deca2e419e07c00889f0984 and follow-up commits. This was raised several times on chat and calls and simplifies constructor APIs and is a direct preparation for the `eips` option addition since this would spin-up a too-complex-to-handle possibility matrix in constructors regarding the chain setup
- The default HF setting is now centralized in Common and static references in consuming libraries (VM, tx, block, blockchain) have been removed 2ba57014c889f0c5e4d086e7764e86ac7f2b6e9b, this solves #822
- There is a new `Common.setHardforkByBlockNumber()` method. This is used by the `Block` library to allow for a new `hardforkByBlockNumber` option for blocks and headers which trigger a HF adoption by the block number set to the block, this solves the tx initialization problems we had along 116ea781f4df59fc452a25d672d6aabadba667bc, see #757. Since there were some immutability discussion (which we won't completely solve on this first round) along I also made the `setGenesisParams()` function private and switched to a constructor option `initWithGenesisHeader` to initialize a genesis header 623dbe0f5ceac83352d4fbef250d12b1b4e7314c, which was one of the major inconsistency creators as one step towards an immutable `Block` library and also added a "Please add header data only along instantiation" warning in the constructor docs.

Sorry that this got so complex but actually all of the changes from above got triggered by follow-up demands of previous changes so that things got consistent again. Hope it was worth the effort and we get this merged. 😀  If you see improvements to be done feel free - with a short note on Discord maybe - to directly work on this and push here.

jochem-brouwer commented:
Wow this is a heavy one - but the description of these changes sound all super good to me! I don't think it is efficient if I start looking at why tests fail at this point but let me know if I can help over there. I will quickly glance over this code and comment on it.

holgerd77 commented:
@jochem-brouwer thanks! 😄  I think one think which would be the most effective would be if you do a commit-by-commit review - I won't do any rebases here any more or something and this will remain stable - and if you then explicitly confirm which commits have been reviewed.

holgerd77 commented:
@jochem-brouwer pushed again and fixed (hopefully, we'll see on CI) VM API tests failing. Selected state tests are still failing if you want to have a look there once time permits that would be extremely helpful.

jochem-brouwer commented:
State tests fail to run:

```
Error: Chain/hardfork options are not allowed any more on initialization
    at new Transaction (/home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/tx/src/transaction.ts:75:13)
    at exports.makeTx (/home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/vm/tests/util.js:92:12)
    at runTestCase (/home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/vm/tests/GeneralStateTestsRunner.js:61:12)
    at runStateTest (/home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/vm/tests/GeneralStateTestsRunner.js:133:7)
    at /home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/vm/tests/tester.js:112:13
    at fileCallback (/home/runner/work/ethereumjs-vm/ethereumjs-vm/packages/vm/tests/testLoader.js:56:11)
```

Looks like the state test runner still uses the chain/hf options in the constructor.

ryanio commented on packages/block/src/header.ts:
```diff
@@ -133,6 +140,14 @@ export class BlockHeader {
     ]
     defineProperties(this, fields, data)
 
+    if (options.hardforkByBlockNumer) {
```
```suggestion
    if (options.hardforkByBlockNumber) {
```

ryanio commented on packages/block/src/types.ts:
```diff
@@ -7,21 +7,31 @@ import { Block } from './block'
    */
-  chain?: number | string
-
+  hardforkByBlockNumer?: boolean
```
```suggestion
  hardforkByBlockNumber?: boolean
```

ryanio commented on packages/block/test/difficulty.spec.ts:
```diff
@@ -67,15 +69,25 @@ tape('[Header]: difficulty tests', (t) => {
+          uncleHash: test.parentUncles,
+        },
+      }
+      const parentBlock = new Block(parentData, { common, hardforkByBlockNumer: true })
```
```suggestion
      const parentBlock = new Block(parentData, { common, hardforkByBlockNumber: true })
```

ryanio commented on packages/block/test/difficulty.spec.ts:
```diff
@@ -67,15 +69,25 @@ tape('[Header]: difficulty tests', (t) => {
+          number: test.currentBlockNumber,
+        },
+      }
+      const block = new Block(blockData, { common, hardforkByBlockNumer: true })
```
```suggestion
      const block = new Block(blockData, { common, hardforkByBlockNumber: true })
```

jochem-brouwer commented:
Looked over the commits, have to do a more thorough review later.

holgerd77 commented:
Rebased this, had a final more distant look and I think I am now confident on all changes. Did some last fixes and modifications - with the most notable being DEFAULT_HARDFORK moved inside the Common class as a readonly param to be accessible from a Common instance.

Tests are now passing, the `block` tests can be safely ignored, these randomly pass (e.g. on the run before) or fail. This will likely be fixed along #861 

So this should be ready to review now. 😄

ryanio commented on packages/common/src/index.ts:
```diff
@@ -15,7 +35,9 @@ interface hardforkOptions {
  */
 export default class Common {
-  private _hardfork: string | null
+  readonly DEFAULT_HARDFORK: string = 'petersburg'
```
closes #822

ryanio commented on packages/vm/CHANGELOG.md:
````diff
@@ -40,13 +40,21 @@ The following HFs have been added:
 ```
 
+**Breaking**: Please note that the options to directly pass in
+`chain` and `hardfork` strings have been removed to simpify the API.
````
```suggestion
`chain` and `hardfork` strings have been removed to simplify the API.
```

ryanio reviewed (approved):
love it, great PR and nice improvements!

i saw updates to the changelogs for blockchain and vm, but not for:
* block
  * constructor opts
  * new opts: hardforkByBlockNumber, initwithGenesisHeader
* tx
  * constructor opts (i made a note and can include this in my tx refactor pr since I set up the unreleased changelog notes in there)
* common
  * setHardforkForBlockNumber

This PR is big enough as it is so I will merge in and we can handle the above separately :)

(also need to regenerate the docs, but that can be done closer to v5 release prep)
